### PR TITLE
Filter TMF messages for security

### PR DESCRIPTION
**Pull request by bukepo**
According to spec 1.1, section 10.6.2
> A TMF CoAP message sent to the Thread Management UDP port (:MM) MUST comply with at least one of these addressing rules:
> * The IPv6 source address is an RLOC or ALOC.
> * The IPv6 destination address is an RLOC or ALOC.
> * The IPv6 destination address is a Link-Local address.

This PR,

* adds a general interceptor feature to CoAP server.
* filters out messages not complying with the addressing rules of TMF.

### CODE REVIEW

File: src/core/thread/thread_netif.cpp [hunk 1]
```diff
@@ -171,6 +172,23 @@ ThreadError ThreadNetif::RouteLookup(const Ip6::Address &source, const Ip6::Addr
...
+ThreadError ThreadNetif::TmfFilter(const otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    const Ip6::MessageInfo &messageInfo = *static_cast<const Ip6::MessageInfo *>(aMessageInfo);
+    ThreadError error = kThreadError_None;
+
+    VerifyOrExit(messageInfo.GetSockAddr().IsRoutingLocator() ||
+                 messageInfo.GetPeerAddr().IsRoutingLocator() ||
+                 messageInfo.GetSockAddr().IsLinkLocal() ||
+                 messageInfo.GetSockAddr().IsAnycastRoutingLocator() ||
+                 messageInfo.GetPeerAddr().IsAnycastRoutingLocator(),
```
Comment (librasungirl) [inline src/core/thread/thread_netif.cpp]: If one address is` AnycastRoutingLocator`,  it must be `RoutingLocator`.
So here `IsAnycastRoutingLocator()` conditions would never have the chance to be executed.

Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: I'm not sure if an ALOC is a RLOC. From the spec, it seems to be true. What do you think? @jwhui

Comment (librasungirl) [inline src/core/thread/thread_netif.cpp]: Sorry if my previous comment confused you. 
What I want to say is ALOC(`IID: 0000:00FF:FE00:FCxx`) shares the similar address structure as RLOC(`IID:0000:00FF:FE00:xxxx`). For an ALOC address, `IsRoutingLocator()` should always be true, thus no need for the condition` IsAnycastRoutingLocator() `.

Comment (jwhui) [inline src/core/thread/thread_netif.cpp]: I agree that the current implementation of `IsRoutingLocator()` makes the `IsAnycastRoutingLocator()` check redundant.  However, we may choose to make `IsRoutingLocator()` more strict in the future.

It is a minor nit, but I agree that ordering the code to match the spec would be nicer.

Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: Is it better to order the code by the probability? @jwhui

Comment (jwhui) [inline src/core/thread/thread_netif.cpp]: I like the thinking.  However, I still think we need to align the logic better with the specification.

The logic above allows a message to pass if any of the above conditions are true.  For example, the spec says that the source address must be an RLOC or ALOC.  However, the logic above will allow any source address as long as the destination address matches one of the above.

File: src/core/coap/coap_server.hpp [hunk 2]
```diff
@@ -288,6 +288,17 @@ class Server : public CoapBase
 {
 public:
     /**
+     * This function pointer is called before CoAP server processing a CoAP packets.
+     * @param[in]   aMessage        A pointer to the message.
+     @ @param[in]   aMessageInfo    A pointer to the message info associated with @p aMessage.
+     *
+     * @retval  kThreadError_None   allow server to continue processing, otherwise the message
+     *                              should be dropped.
+     *
```
Comment (jwhui) [inline src/core/coap/coap_server.hpp]: nit: document possible return values

File: src/core/thread/thread_netif.cpp (hunk 1 above)
Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: It seems the spec means if any one condition matches, the message can pass. Did I miss anything?

File: src/core/coap/coap_server.hpp (hunk 2 above)
Comment (bukepo) [inline src/core/coap/coap_server.hpp]: added.

File: src/core/thread/thread_netif.cpp (hunk 1 above)
Comment (jwhui) [inline src/core/thread/thread_netif.cpp]: Reading the spec again, you are correct.  At the same time, I think the spec should be updated to be more strict and say both of the following rules must match:

- Source Address is RLOC, ALOC, or link-local address.
- Destination Address is RLOC, ALOC, or link-local address.

In other words, we should not allow messages to pass that have global IPv6 source address with RLOC destination address, for example.

File: src/core/thread/thread_netif.cpp [hunk 3]
```diff
@@ -171,6 +172,28 @@ ThreadError ThreadNetif::RouteLookup(const Ip6::Address &source, const Ip6::Addr
     return error;
 }
 
+ThreadError ThreadNetif::TmfFilter(const otMessage *aMessage, const otMessageInfo *aMessageInfo)
```
Comment (jwhui) [inline src/core/thread/thread_netif.cpp]: Pass-by-reference for both arguments?

File: src/core/coap/coap_server.hpp [hunk 4]
```diff
@@ -288,6 +288,19 @@ class Server : public CoapBase
 {
 public:
     /**
+     * This function pointer is called before CoAP server processing a CoAP packets.
+     * @param[in]   aMessage        A pointer to the message.
```
Comment (jwhui) [inline src/core/coap/coap_server.hpp]: nit: add empty line between brief and param docs.

File: src/core/thread/thread_netif.cpp (hunk 3 above)
Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: Do you think that we should expose this interceptor feature to user for the application CoAP server?

Comment (jwhui) [inline src/core/thread/thread_netif.cpp]: If we want to make this a more general solution, then I think we should move this kind of "fire walling" behavior down to the IPv6 layer.  I'm fine with the current implementation, but would not want to project this through the CoAP API.

Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: I also wavered over the layer of this functionality before choosing the current implementation.  I strongly support the general "fire walling" feature. I've changed the way to pass both arguments.

**Pull request merged**